modules/ballerina-http/src/main/java/org/ballerinalang/net/http/MultipartUtil.java

```diff
@@ -0,0 +1,161 @@
+
+/**
+ * Utility functions to handle multipart data.
+ */
+public class MultipartUtil {
+    private static final Logger log = LoggerFactory.getLogger(MultipartUtil.class);
+
+    private static final int ACTUAL_CONTENT = 0;
+    private static final int FILE_NAME = 0;
+    private static final int PART_NAME = 1;
+    private static final int CONTENT_TYPE = 2;
+    private static final int PART_SIZE = 0;
+
+    /**
+     * Extract multipart data from input stream.
+     *
+     * @param inputStream InputStream
+     * @return List<HttpBodyPart>
+     */
+    public static List<HttpBodyPart> extractMultiparts(InputStream inputStream) {
+        List<HttpBodyPart> multiparts = null;
+        try {
+            ObjectInputStream objectInputStream = new ObjectInputStream(inputStream);
+            multiparts = (List<HttpBodyPart>) objectInputStream.readObject();
+        } catch (IOException e) {
+            log.error("IOException occurred while extracting multiparts from recieved inputstream", e);
+        } catch (ClassNotFoundException e) {
+            log.error("ClassNotFoundException occurred while extracting multiparts from recieved inputstream", e);
+        }
+        return multiparts;
+    }
+
+    /**
+     * Build an array of 'Part' structs.
+     *
+     * @param context    Context
+     * @param multiparts list of multipart bodies.
+     * @return BRefValueArray
+     */
+    public static BRefValueArray fillPartsArray(Context context, List<HttpBodyPart> multiparts) {
+        ArrayList<BStruct> parts = new ArrayList<>();
+        for (HttpBodyPart httpBodyPart : multiparts) {
+            BStruct partStruct = ConnectorUtils
+                    .createAndGetStruct(context, Constants.PROTOCOL_PACKAGE_HTTP, Constants.PART);
+            populatePartStructs(httpBodyPart, partStruct);
+            parts.add(partStruct);
+        }
+        if (!parts.isEmpty()) {
+            BStructType type = parts.get(0).getType();
+            BStruct[] result = parts.toArray(new BStruct[parts.size()]);
+            BRefValueArray partsArray = new BRefValueArray(result, type);
+            return partsArray;
+        }
+        return null;
+    }
+
+    /**
+     * Populate 'Part' struct from a http body part.
+     *
+     * @param httpBodyPart Http body part
+     * @param balPart      BStruct
+     */
+    private static void populatePartStructs(HttpBodyPart httpBodyPart, BStruct balPart) {
+        balPart.setBlobField(ACTUAL_CONTENT, httpBodyPart.getContent());
+        balPart.setStringField(FILE_NAME, httpBodyPart.getFileName());
+        balPart.setStringField(PART_NAME, httpBodyPart.getPartName());
+        balPart.setStringField(CONTENT_TYPE, httpBodyPart.getContentType());
+        balPart.setIntField(PART_SIZE, httpBodyPart.getSize());
+    }
+
+    /**
+     * A static class to represent a list of body part as a message data source.
+     */
+    public static class MultipartDataSource implements MessageDataSource {
```

Comment:
I think you can extend `BallerinaMessageDataSource` class, and override only the required methods.